features/22-email-verification/api.md

```diff
@@ -0,0 +1,195 @@
+having custom user data can provide their own provider. The provider is
+extensible in python cloud code only.
+
+### Verification Link/Code Persistence and Consumption
+
+Each verification link and code is to be randomly generated and persisted to
+the database. Each code has an associated expiry time
+and can only be used once. If there are multiple verification links and codes
+generated for the same user, all of them should be usable.
+
+## Sample usage
+
+### To verify user data:
+
+1. User click on the link, enter code.
+2. Email address or phone number is marked as verified.
+3. If all required user data field is verified, the user is marked as verified.
+4. The App can check verification status through API.
+
+### When signing up
+
+1. User enter email address and phone number through signup API.
+2. If server is set to automatically send verify link/code, verification
+   email/SMS is sent. If not, app can request verification through API.
+3. Verify user data (see above).
+
+### When updating
+
+1. User update email address or phone number through record API.
+2. The email address or phone number is marked as unverified.
+3. If some required user data field is unverified, the user is marked as
+   unverified.
+4. If server is set to automatically send verify link/code, verification
+   email/SMS is sent. If not, app can request verification through API.
+5. Verify user data (see above).
+
+### If user verified flag is not automatically updated.
+
+1. User update email address or phone number through record API.
+2. The email address or phone number is marked as unverified.
+3. User verified flag is not updated.
+4. Developer write cloud function to toggle the user verified flag through API.
+
+## Sample APIs
+
+When signing up:
+
+```javascript
+// email
+skygear.auth.signupWithEmail("johndoe@example.com", "password")
+.then((user) => {
+  console.log(skygear.auth.verified); // false
+});
+```
+
+To check if the user is verified:
+
+```javascript
+skygear.auth.whoAmI()
+.then((user) => {
+  console.log(skygear.auth.verified); // false
+});
+```
+
+The app can request verification again:
+
+```javascript
+// email
+skygear.auth.requestVerification('email')
+.then(() => {
+  // email sent
+});
+```
+
+The app can request verification on other fields of user table:
+
+```javascript
+// email
+skygear.auth.requestVerification('emergency_phone')
+.then(() => {
+  // SMS sent
+});
+```
+
+To check for verification:
+
+```javascript
+// verify email 
+skygear.auth.verifyWithCode('email', '31DF2310FAA1')
+.then((user) => {
+  console.log(skygear.auth.verified); // true
+  console.log(user.email_verified); // true
+});
+```
+
+If verification is required and the user is not verified:
+
+```javascript
+skygear.someAction()
+.then(() => {
+  // the promise is rejected and this callback is not called
+}, (err) => {
+  console.log(err.name); // 'VerificationRequired'
+});
+```
+
```

Comment:
I think we also should include (could be done separately later) the spec for cloud functions:

1. Support verified_required as a parameter for lambda functions, so users can limit lambda functions to verified users easily.

2. Support `after_user_verified` as a call back for cloud functions when user is verified.

Reply:
It doesn't meant we have to implement than rightaway, but I guess that would be more complete for the verification feature for Skygear